[Build Speed] Reduce includes of StyleBasicShape.h

jernoble opened the pull request:
#### f9db93196fba762fe33ab135f6704b14bcc6b93c
<pre>
[Build Speed] Reduce includes of StyleBasicShape.h
<a href="https://rdar.apple.com/163812454">rdar://163812454</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=301767">https://bugs.webkit.org/show_bug.cgi?id=301767</a>

Reviewed by Sam Weinig.

Before this patch, StyleBasicShape.h was the 4th most expensive header in a
Unified WebCore build. It was included 310 times, at an average cost of 1121
ms, for a total CPU cost of 5m47s.

It is included in 4 headers total, and in only one of those can its full
definitions be replaced with forward declarations. That forward declaration is
complex enough to warrant its own header file, StyleShapeForward.h.

After this patch, StyleBasicShape.h is the 24th most expensive header. It is
included 210 times for a total CPU cost of 1m37s.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/TextBoxPainter.h:
* Source/WebCore/rendering/shapes/LayoutShape.h:
* Source/WebCore/style/values/shapes/StyleShapeForward.h: Added.

Canonical link: <a href="https://commits.webkit.org/302486@main">https://commits.webkit.org/302486@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f5d4e35d2728987ef23c158433e34973e63195d4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/129026 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/1285 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/39862 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/136407 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/80395 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/20b00818-e34f-4424-8cd7-8ba470cd50a5) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/130897 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/1217 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/1162 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/66133 "") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/7900e271-2858-49ca-93de-cff7ac21a626) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/131973 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/943 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/115578 "Passed tests") | [⏳ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/API-Tests-WPE-EWS "Waiting to run tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3b9b74e3-6fa9-4082-843e-8e7da0267560) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/871 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/33693 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/79686 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/109311 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/34191 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/138881 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/1080 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/106779 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/1137 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/111914 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/106605 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27188 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/892 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/30438 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/53585 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/1154 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/64494 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/988 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/1037 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/1081 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/style/values/shapes/StyleShapeForward.h:
```diff
@@ -0,0 +1,62 @@
+using PolygonFunction = FunctionNotation<CSSValuePolygon, Polygon>;
+using ShapeFunction = FunctionNotation<CSSValueShape, Shape>;
+
+using BasicShape = Variant<
```
I think a simpler approach would probably be to convert BasicShape to a struct wrapping a variant. That's a bit more of an involved, change, but I would be happy to do it.

jernoble replied:
The goal with this patch was to reduce the build time without making functional changes.

weinig commented on Source/WebCore/rendering/RenderBox.h:
```diff
@@ -23,6 +23,7 @@
 
 #pragma once
 
+#include <WebCore/CSSPrimitiveNumeric.h>
```
This is a pretty heavy include, so if it can be avoided here, that would be good. What needed it?

jernoble replied:
It was previously included via StyleBasicShape.h, so it's still a net decrease in build time.

jernoble replied:
Without this, you get:
```
Source/WebCore/rendering/RenderBox.h:429:98: error: no template named 'LengthPercentage' in namespace 'WebCore::CSS'; did you mean 'Style::LengthPercentage'?
  429 |     std::optional<LayoutUnit> computePercentageLogicalHeight(const Style::UnevaluatedCalculation<CSS::LengthPercentage<CSS::NonnegativeUnzoomed, float>>& logicalHeight, UpdatePercentageHeightDescendants = UpdatePercentageHeightDescendants::Yes) const;
```
While there's a forward declaration for `Style::LengthPercentage` in `StylePrimitiveNumeric+forward.h`, there isn't a convenient forward declaration of `CSS::LengthPercentage`.

weinig commented on Source/WebCore/rendering/TextBoxPainter.h:
```diff
@@ -30,6 +30,7 @@
 #include "InlineIteratorTextBox.h"
 #include "PaintInfo.h"
 #include "RenderObject.h"
+#include "StylePrimitiveNumeric.h"
```
Is this for `TextBoxPainter::rotateShadowOffset`? Does that really need to be inline?

jernoble replied:
I'm deliberately trying _not_ to make decisions about whether there is a perf benefit to having a function be inline. I'm only moving things out-of-line when the method is private or virtual.

jernoble replied:
And yes, this is needed for `rotateShadowOffset`.

jernoble committed:
[Build Speed] Reduce includes of StyleBasicShape.h
rdar://163812454
https://bugs.webkit.org/show_bug.cgi?id=301767

Reviewed by Sam Weinig.

Before this patch, StyleBasicShape.h was the 4th most expensive header in a
Unified WebCore build. It was included 310 times, at an average cost of 1121
ms, for a total CPU cost of 5m47s.

It is included in 4 headers total, and in only one of those can its full
definitions be replaced with forward declarations. That forward declaration is
complex enough to warrant its own header file, StyleShapeForward.h.

After this patch, StyleBasicShape.h is the 24th most expensive header. It is
included 210 times for a total CPU cost of 1m37s.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderBox.h:
* Source/WebCore/rendering/TextBoxPainter.h:
* Source/WebCore/rendering/shapes/LayoutShape.h:
* Source/WebCore/style/values/shapes/StyleShapeForward.h: Added.

Canonical link: https://commits.webkit.org/302486@main
